Gemini -(Kumo) Task created in a session that has been invalidated

gnkeshavamurthy opened the pull request:
# DESCRIPTION
Fixing `Task created in a session that has been invalidated` crash


# CHANGE
- Changed Service from `class` to `actor`
- Updated `perform()` with async & await

gnkeshavamurthy committed:
Gemini -(Kumo) Task created in a session that has been invalidated

gnkeshavamurthy committed:
reverted invalidations change

gnkeshavamurthy committed:
updated min support macOS to 12

powerje commented on Sources/Kumo/Blobs/BlobCache.swift:
```diff
@@ -195,4 +195,18 @@ public class BlobCache {
         persistentStorage.clean()
     }
 
+    private func fetchImage(from url: URL) -> AnyPublisher<URL, Error> {
```
This could be any file, not just images, maybe just "fetch" is fine here?

powerje commented on Tests/KumoTests/Fixtures/Common/NetworkTest.swift:
```diff
@@ -52,9 +52,40 @@ class NetworkTest: XCTestCase {
+        }.eraseToAnyPublisher()
+    }
+    
+    func performs<Method: _RequestMethod, Resource: _RequestResource, Parameters: _RequestParameters, Body: _RequestBody, Key: _ResponseNestedKey>(_ request: HTTP
```
I'm confused by the naming between `perform` vs `performs` and when I'd use one vs the other?

powerje commented on Tests/KumoTests/Fixtures/Services/GetTests.swift:
```diff
@@ -5,47 +5,47 @@ import XCTest
     
     func testSuccessfulGetRequestEmittingVoid() {
-        successfulTest(of: service.perform(HTTP.Request.get("get")))
+        successfulTest(of: performs(HTTP.Request.get("get")))
```
Is this still calling a function on the `service` somehow?

powerje reviewed (commented):
Had some questions about the changes

powerje commented:
Running the tests (`$ swift test`) there are some warnings:
```
Service.swift:285:10: note: calls to instance method 'downloadResultToURL(url:response:)' from outside of its actor context are implicitly asynchronous
    func downloadResultToURL(url: URL?, response: URLResponse?) throws -> URL {
...
 warning: non-sendable type 'OperationQueue?' passed in call to nonisolated initializer 'init(baseURL:runsInBackground:logger:delegateQueue:configuration:)' cannot cross actor boundary
    let service = Service(baseURL: URL(string: "https://httpbin.org")!, logger: TestLogger())
...
Swift.Optional:1:21: note: generic enum 'Optional' does not conform to the 'Sendable' protocol
@frozen public enum Optional<Wrapped> : ExpressibleByNilLiteral {
```

I think we should address these?

gnkeshavamurthy committed:
Fixed review comments

gnkeshavamurthy committed:
Updated unit tests

gnkeshavamurthy committed:
Fixed warning

gnkeshavamurthy commented:
> Running the tests (`$ swift test`) there are some warnings:
> 
> ```
> Service.swift:285:10: note: calls to instance method 'downloadResultToURL(url:response:)' from outside of its actor context are implicitly asynchronous
>     func downloadResultToURL(url: URL?, response: URLResponse?) throws -> URL {
> ...
>  warning: non-sendable type 'OperationQueue?' passed in call to nonisolated initializer 'init(baseURL:runsInBackground:logger:delegateQueue:configuration:)' cannot cross actor boundary
>     let service = Service(baseURL: URL(string: "https://httpbin.org")!, logger: TestLogger())
> ...
> Swift.Optional:1:21: note: generic enum 'Optional' does not conform to the 'Sendable' protocol
> @frozen public enum Optional<Wrapped> : ExpressibleByNilLiteral {
> ```
> 
> I think we should address these?

Fixed this warning

gnkeshavamurthy committed:
Updated osx & tvos deployement targets to 12 & 15

powerje committed:
Updated podspec for 3.0.0.

Updated swift version and removed `RxSwift` dependency from podspec.
`RxSwift` requirement was removed in the 2.0.0 update but stuck around
the podspec erroneously.
